py-data-manipulation.Rmd

```diff
@@ -0,0 +1,457 @@
+error messages if it can't run your code that will help you figure out where
+the problem is.
+
+```
+# this code generates a syntax error
+dog_licenses['animal_name'
+```
+
+The very bottom of the error message is the most helpul place to start looking:
+this says what kind of error it is (e.g. `SyntaxError`, `ValueError`,
+`TypeError`, etc) and sometimes includes a message that describes what went
+wrong.
+
+It also points to the place where the error happened: in this case, there's a
+missing square bracket, so the error we got was that Python reached the end of
+the file before it expected to because we didn't include proper punctuation.
+
+```python
+# this code generates a runtime error (there's nothing wrong with the syntax,
+# but Python can't execute it because of a different error)
+
+dog_licenses['animal_name'] < dog_licenses['zip_code']
+```
+
+The last line again tells us what went wrong, but this time we get a
+*traceback*: Python is showing us the series of steps that it followed
+internally before it hit an error. This is helpful because often the error will
+ultimately occur inside a function we're using but whose code we haven't ever
+seen, so Python shows us both what specifically went wrong inside the function,
+but also what line of our code caused the original cascade. Here, the offending
+line is pointed to at the very top of the traceback, and by reading the error
+message at the bottom we can see that the problem was that we can't use the
+'less than' `<` comparison between two objects when one is a string
+(`dog_licenses['animal_name']`) and the other is a float (`dog_licenses['zip_code']`).
+
+## How can I operate on subsets of my data?
+
+Many data analysis tasks can be approached using the *split-apply-combine
+paradigm*: split the data into groups, apply some analysis to each group, and
+then combine the results.
+
+`pandas` facilitates this workflow through the use of `groupby()` to split
+data and summary/aggregation functions such as `mean()` which collapses each
+group into a single-row summary of that group. The arguments to `groupby()` are
+the column names that contain the categorical variables by which summary
+statistics should be calculated.
+
+Let's start with a smaller data frame of the top 10 most common dog breeds and
+with only boroughs that are listed 5 or more times
+
+```python
+# get a list of the 10 most common breed names
+common_breeds = dog_licenses_large_boroughs['breed_name'].value_counts()[:10].index
+
+# create a new data frame with only the common breeds
+dog_licenses_common_breeds = dog_licenses_large_boroughs[dog_licenses_large_boroughs['breed_name'].isin(common_breeds)]
+```
+
+`groupby()` enables us to break the data down into groups and perform operations
+on each group. To look at how many animals of each sex there are in each of the
+most common breeds, we can use `groupby()`:
+
+```python
+dog_licenses_common_breeds.groupby('breed_name')['animal_gender'].value_counts()
```

Comment:
There is also `groupby().size()` but I maybe we shuold stick to value_counts since we already introduced it.